index.js

```diff
@@ -1,29 +1,92 @@
-    }
-    */
-
-    /*
-    Declaration (decl, postcss) {
-      // The faster way to find Declaration node
-    }
-    */
-
-    /*
-    Declaration: {
-      color: (decl, postcss) {
-        // The fastest way find Declaration node if you know property name
-      }
-    }
-    */
+module.exports = validateCustomProps;
+module.exports.postcss = true;
+
+/**
+ * Object to define an incorrect CSS custom property
+ *
+ * @typedef {Object} WrongProp
+ * @property {string} path - absolute path of the CSS file
+ * @property {number} line - line number containing the custom property
+ * @property {string} key - key of the custom property
+ * @property {string} current - current value of the custom property
+ * @property {string} expexted - expected correct value of the custom property
+ */
+
+/**
+ * Callback to handle the response of the plugin
+ *
+ * @callback requestCallback
+ * @param {Object} error - an error object
+ * @param {WrongProp[]} data - a list of incorrect custom properties
+ * @returns
+ */
+
+/**
+ * Postcss plugin to validate fallback value of CSS custom properties
+ *
+ * @param {Object} opts - plugin options
+ * @param {Object} opts.properties - CSS custom properties to validate values from
+ * @param {boolean} opts.write [write=false] - update the file with updated CSS
+ * @param {requestCallback} opts.callback [callback=() => {}] - callback to handle the response
+ * @returns Object
+ */
+function validateCustomProps({ properties, write = false, callback = noop }) {
+  if (!properties) {
+    throw new Error("No properties present in options");
   }
+
+  return {
+    postcssPlugin: "postcss-customprop-validate",
+    prepare(result) {
+      const wrongProps = [];
+
+      return {
+        Declaration(decl) {
+          const customProp = getCustomProp(decl);
+
+          // exit if the declaration is not a custom property
+          if (!customProp) return;
+
+          const { key, value } = customProp;
+          const expected = properties[key];
+
+          // exit if a custom property does not exist in `properties` plugin options
+          if (!expected) return;
+
+          // exit if the fallback value of the custom property is correct
+          if (value === expected) return;
+
+          wrongProps.push({
+            path: decl.source.input.file,
+            line: decl.source.start.line,
+            key,
+            current: value,
+            expected,
+          });
+
+          decl.replaceWith({
+            prop: decl.prop,
+            raws: decl.raws,
+            value: `var(${key}, ${expected})`,
+          });
+        },
+        OnceExit(root) {
+          const resultant = root.toResult().css;
+          const { file } = root.source.input;
+
+          if (write && wrongProps.length) {
+            writeFile(file, resultant, (err) => callback(err, wrongProps));
+          } else {
+            callback(null, wrongProps);
+          }
+
+          result.wrongProps = wrongProps;
```

Comment:
@abhinavpreetu does this need to be the last statement? seems to me that the `writeFile` or the `callback` functions are not using `result`, so should `result.wrongProps = wrongProps` happen earlier?

Reply:
Yeah, the result is what the plugin produces after processing the CSS. The users will have access to result anyways when they run the plugin. For our use case, we are just concerned with `wrongProps` and this is what we give the users as the 2nd argument of the callback. 
Now, I also wanted the users to get the `wrongProps` in `result`, hence I append it. Now, I can place it anywhere in the `onceExit()` function and it won't make any difference.

Reply:
okay. does it make sense to move it a bit upwards then? that ways, either the call to `writeFile` or `callback` becomes the last thing happening to the function and hence this question doesn't come to mind.

Reply:
Understood. Makes sense